### PR TITLE
Use pivot reports to fetch audiences metrics.

### PR DESCRIPTION
**Pull request by ankitrox**
## Summary

Addresses issue:

- #8726 

## PR Author Checklist

- [ ] My code is tested and passes existing unit tests.
- [ ] My code has an appropriate set of unit tests which all pass.
- [ ] My code is backward-compatible with WordPress 5.2 and PHP 7.4.
- [ ] My code follows the [WordPress](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) coding standards.
- [ ] My code has proper inline documentation.
- [ ] I have added a QA Brief on the issue linked above.
- [x] I have signed the Contributor License Agreement (see <https://cla.developers.google.com/>).

---------------

_Do not alter or remove anything below. The following sections will be managed by moderators only._

## Code Reviewer Checklist

- [x] Run the code.
- [x] Ensure the acceptance criteria are satisfied.
- [x] Reassess the implementation with the IB.
- [x] Ensure no unrelated changes are included.
- [x] Ensure CI checks pass.
- [x] Check Storybook where applicable.
- [x] Ensure there is a QA Brief.

## Merge Reviewer Checklist

- [x] Ensure the PR has the correct target branch.
- [x] Double-check that the PR is okay to be merged.
- [x] Ensure the corresponding issue has a ZenHub release assigned.
- [x] Add a changelog message to the issue.

### CODE REVIEW

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js:1]: One of the points of the AC/IB has been missed here, which is that all the destructuring/parsing of report rows should be removed from this component, and instead the relevant report rows for an audience should be passed directly to the `AudienceTile` component, which will need a new prop structure and associated logic.

Note the relevant points in the AC and IB:

AC:
> * The `AudienceTiles` and `AudienceTile` components should be further refactored so the reports that are retrieved in `AudienceTiles` are passed directly to `AudienceTile`, and the corresponding data parsing/extraction logic is moved from `AudienceTiles` to `AudienceTile`. Hopefully this logic can be simplified in the process.

IB:
>  * [ ] Update `AudienceTile` to use the new prop structure defined above.
>  * [ ] Update `AudienceTile` stories to use this new pivot report data-mock functionality defined in #8484

I.e. these areas of the code are up for a refactor in this issue:

https://github.com/google/site-kit-wp/blob/3a0ed6d1837eeb48d383704a928f81cbb3028b47/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js#L389-L462

https://github.com/google/site-kit-wp/blob/3a0ed6d1837eeb48d383704a928f81cbb3028b47/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js#L482-L540

https://github.com/google/site-kit-wp/blob/3a0ed6d1837eeb48d383704a928f81cbb3028b47/assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTile/index.js#L59-L80

Please note, the comments that refer to #8484 were written before #8726 was split off from #8484. They should be taken to refer to this issue instead.

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @ankitrox, good work so far. I've left a number of comments to address, please take a look.

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js [hunk 1]
```diff
@@ -277,6 +304,31 @@ function AudienceTilesWidget( { Widget } ) {
 		}
 	} );
 
+	/**
+	 * Collects the rows of interest for the relevant audience from report.
```
[the remark this answers is not in the document]
Comment (ankitrox) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js:308]: Removed the doc.

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js [hunk 2]
```diff
@@ -131,85 +132,111 @@ function AudienceTilesWidget( { Widget } ) {
...
+				],
+				limit: 3,
+			},
 			{
-				metric: {
-					metricName: 'totalUsers',
-				},
-				desc: true,
+				fieldNames: [ 'audienceResourceName' ],
+				limit: configuredAudiences.length,
```
[the remark this answers is not in the document]
Comment (ankitrox) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js:150]: @techanvil Yes, this limit is required. We have added [validations](https://github.com/google/site-kit-wp/blob/develop/assets/js/modules/analytics-4/utils/validation.js#L291) in #8484 to add `fieldNames` and `limit`.

[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/components/audience-segmentation/dashboard/AudienceTilesWidget.js:150]: Thanks for clarfying! For some reason I thought it wasn't a required parameter, but I see it is as per the API [docs](https://developers.google.com/analytics/devguides/reporting/data/v1/rest/v1beta/Pivot).

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 3]
```diff
@@ -862,46 +905,57 @@ export function getAnalytics4MockPivotResponse( options ) {
...
-							ANALYTICS_4_DIMENSION_OPTIONS[ dimension ]( i );
-						if ( dimensionValue ) {
-							observer.next( dimensionValue );
-						} else {
-							break;
+		fieldNames.forEach( ( dimension ) => {
+			// If an array of filter values is provided for the dimension, use that to generate the stream.
+			if (
+				dimension &&
+				args?.dimensionFilters?.[ dimension ]?.some(
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:913]: Hey @ankitrox I just realised `every()` would be more appropriate here:

```suggestion
				args?.dimensionFilters?.[ dimension ]?.every(
```

File: assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json [hunk 4]
```diff
@@ -0,0 +1,437 @@
...
+              {
+                "value": "India"
+              }
+            ]
+          }
+        ],
+        "rowCount": 3
+      }
+    ],
+    "aggregates": [
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json:122]: The number and shape of the aggregates is also not correct. As you can see, for each pivot, we should have effectively have the number of rows (i.e. `limit` in our output) x 3 (TOTAL, MINIMUM and MAXIMUM).

So there should be 2x3 + 3x3 + 3x3 = 24 entries in `aggregates`. As you can see the current output only has 9 entries. Also, the each entry should contain 4 dimension values.

<details>
<summary>Expected shape (taken from my example)</summary>

```json
	"aggregates": [
		{
			"dimensionValues": [
				{
					"value": "new"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				}
			],
			"metricValues": [
				{
					"value": "4892"
				},
				{
					"value": "4892"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "iOS"
				}
			],
			"metricValues": [
				{
					"value": "2104"
				},
				{
					"value": "2104"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "Android"
				}
			],
			"metricValues": [
				{
					"value": "1936"
				},
				{
					"value": "1936"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "Dublin"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_TOTAL"
				}
			],
			"metricValues": [
				{
					"value": "1202"
				},
				{
					"value": "1202"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "Windows"
				}
			],
			"metricValues": [
				{
					"value": "690"
				},
				{
					"value": "690"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "Dublin"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_MAX"
				}
			],
			"metricValues": [
				{
					"value": "551"
				},
				{
					"value": "551"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "iOS"
				}
			],
			"metricValues": [
				{
					"value": "551"
				},
				{
					"value": "551"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "new"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				}
			],
			"metricValues": [
				{
					"value": "551"
				},
				{
					"value": "551"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "returning"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "RESERVED_TOTAL"
				}
			],
			"metricValues": [
				{
					"value": "530"
				},
				{
					"value": "530"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "Android"
				}
			],
			"metricValues": [
				{
					"value": "430"
				},
				{
					"value": "430"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "Cork"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_TOTAL"
				}
			],
			"metricValues": [
				{
					"value": "244"
				},
				{
					"value": "244"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_TOTAL"
				},
				{
					"value": "Sao Paulo"
				},
				{
					"value": "Brazil"
				},
				{
					"value": "RESERVED_TOTAL"
				}
			],
			"metricValues": [
				{
					"value": "196"
				},
				{
					"value": "196"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "Cork"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_MAX"
				}
			],
			"metricValues": [
				{
					"value": "191"
				},
				{
					"value": "191"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "Windows"
				}
			],
			"metricValues": [
				{
					"value": "121"
				},
				{
					"value": "121"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "Sao Paulo"
				},
				{
					"value": "Brazil"
				},
				{
					"value": "RESERVED_MAX"
				}
			],
			"metricValues": [
				{
					"value": "83"
				},
				{
					"value": "83"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "returning"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				},
				{
					"value": "RESERVED_MAX"
				}
			],
			"metricValues": [
				{
					"value": "78"
				},
				{
					"value": "78"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "Dublin"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_MIN"
				}
			],
			"metricValues": [
				{
					"value": "4"
				},
				{
					"value": "4"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "Cork"
				},
				{
					"value": "Ireland"
				},
				{
					"value": "RESERVED_MIN"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "Android"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "Windows"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "iOS"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "Sao Paulo"
				},
				{
					"value": "Brazil"
				},
				{
					"value": "RESERVED_MIN"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "new"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		},
		{
			"dimensionValues": [
				{
					"value": "returning"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				},
				{
					"value": "RESERVED_MIN"
				}
			],
			"metricValues": [
				{
					"value": "1"
				},
				{
					"value": "1"
				}
			]
		}
	],
```

</details>

<details>
<summary>Current output</summary>

```json
    "aggregates": [
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MIN"
          },
          {
            "value": "city"
          }
        ],
        "metricValues": [
          {
            "value": "27"
          },
          {
            "value": "91"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MIN"
          },
          {
            "value": "audienceResourceName"
          }
        ],
        "metricValues": [
          {
            "value": "86"
          },
          {
            "value": "53"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MIN"
          },
          {
            "value": "country"
          }
        ],
        "metricValues": [
          {
            "value": "38"
          },
          {
            "value": "71"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MAX"
          },
          {
            "value": "city"
          }
        ],
        "metricValues": [
          {
            "value": "21"
          },
          {
            "value": "74"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MAX"
          },
          {
            "value": "audienceResourceName"
          }
        ],
        "metricValues": [
          {
            "value": "66"
          },
          {
            "value": "80"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_MAX"
          },
          {
            "value": "country"
          }
        ],
        "metricValues": [
          {
            "value": "16"
          },
          {
            "value": "22"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_TOTAL"
          },
          {
            "value": "city"
          }
        ],
        "metricValues": [
          {
            "value": "45"
          },
          {
            "value": "34"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_TOTAL"
          },
          {
            "value": "audienceResourceName"
          }
        ],
        "metricValues": [
          {
            "value": "23"
          },
          {
            "value": "91"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "RESERVED_TOTAL"
          },
          {
            "value": "country"
          }
        ],
        "metricValues": [
          {
            "value": "38"
          },
          {
            "value": "66"
          }
        ]
      }
    ],
```

File: assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json [hunk 5]
```diff
@@ -0,0 +1,437 @@
...
+          {
+            "value": "38"
+          },
+          {
+            "value": "66"
+          }
+        ]
+      }
+    ],
+    "rows": [
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json:286]: Lastly, the number of `rows` is not as expected either.

The number of rows should be the product of multiplying the row count for each pivot together. So with pivots returning 2, 3 and 3 rows, the report row count should be 2x3x3 = 18.

However the current output only has 6 rows, so it evidently doesn't have the right combination of values.

<details>
<summary>Expected shape (taken from my example)</summary>

```json
"rows": [
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "551"
			},
			{
				"value": "551"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "430"
			},
			{
				"value": "430"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "191"
			},
			{
				"value": "191"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "121"
			},
			{
				"value": "121"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "83"
			},
			{
				"value": "83"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "78"
			},
			{
				"value": "78"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "70"
			},
			{
				"value": "70"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "64"
			},
			{
				"value": "64"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "36"
			},
			{
				"value": "36"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "26"
			},
			{
				"value": "26"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "21"
			},
			{
				"value": "21"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Dublin"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "14"
			},
			{
				"value": "14"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "12"
			},
			{
				"value": "12"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "new"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "8"
			},
			{
				"value": "8"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "iOS"
			}
		],
		"metricValues": [
			{
				"value": "7"
			},
			{
				"value": "7"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Android"
			}
		],
		"metricValues": [
			{
				"value": "4"
			},
			{
				"value": "4"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Sao Paulo"
			},
			{
				"value": "Brazil"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "4"
			},
			{
				"value": "4"
			}
		]
	},
	{
		"dimensionValues": [
			{
				"value": "returning"
			},
			{
				"value": "Cork"
			},
			{
				"value": "Ireland"
			},
			{
				"value": "Windows"
			}
		],
		"metricValues": [
			{
				"value": "1"
			},
			{
				"value": "1"
			}
		]
	}
],
```

<details>
<summary>Current output</summary>

```json
    "rows": [
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/2"
          },
          {
            "value": "Cork"
          },
          {
            "value": "India"
          },
          {
            "value": "India"
          }
        ],
        "metricValues": [
          {
            "value": "98"
          },
          {
            "value": "7"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/2"
          },
          {
            "value": "(not set)"
          },
          {
            "value": "United Kingdom"
          },
          {
            "value": "United Kingdom"
          }
        ],
        "metricValues": [
          {
            "value": "89"
          },
          {
            "value": "62"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/1"
          },
          {
            "value": "Dublin"
          },
          {
            "value": "United States"
          },
          {
            "value": "United States"
          }
        ],
        "metricValues": [
          {
            "value": "55"
          },
          {
            "value": "14"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/1"
          },
          {
            "value": "Cork"
          },
          {
            "value": "India"
          },
          {
            "value": "India"
          }
        ],
        "metricValues": [
          {
            "value": "17"
          },
          {
            "value": "39"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/2"
          },
          {
            "value": "Dublin"
          },
          {
            "value": "United States"
          },
          {
            "value": "United States"
          }
        ],
        "metricValues": [
          {
            "value": "10"
          },
          {
            "value": "73"
          }
        ]
      },
      {
        "dimensionValues": [
          {
            "value": "properties/12345/audiences/1"
          },
          {
            "value": "(not set)"
          },
          {
            "value": "United Kingdom"
          },
          {
            "value": "United Kingdom"
          }
        ],
        "metricValues": [
          {
            "value": "3"
          },
          {
            "value": "13"
          }
        ]
      }
    ],
```

</details>

**Review by techanvil — CHANGES_REQUESTED** [review]
Hi @ankitrox, thanks for the update. This does need another update, though - I've noticed the output for the mock data is not quite as expected, and there are a couple of other points to address. Please take a look.

File: assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json
[the remark this answers is not in the document]
Comment (techanvil) [reply assets/js/modules/analytics-4/utils/__fixtures__/mocked-pivot-report-multi-dimension-fieldnames.json:1]: Btw, before getting stuck into this I'd suggest making the change to replace `country` in the third pivot as per my comment above: https://github.com/google/site-kit-wp/pull/8870#discussion_r1652483235

It should make it a bit easier to debug!

File: assets/js/modules/analytics-4/utils/data-mock.js [hunk 6]
```diff
@@ -949,41 +1005,47 @@ export function getAnalytics4MockPivotResponse( options ) {
...
-								],
-							} )
-						),
-						rowCount: dimensionValues.length,
-					} );
-					return acc;
-				},
-				[]
-			);
+			data.pivotHeaders = options.pivots.map( ( { fieldNames } ) => {
```
Comment (techanvil) [inline assets/js/modules/analytics-4/utils/data-mock.js:1033]: I noticed there was still an issue with `pivotHeaders`.

<details>
<summary>The output was as follows:</summary>

```json 
  "pivotHeaders": [
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "properties/12345/audiences/1"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "properties/12345/audiences/2"
            }
          ]
        }
      ],
      "rowCount": 2
    },
    {
      "pivotDimensionHeaders": [
        {
          "value": "Dublin"
        },
        {
          "value": "(not set)"
        },
        {
          "value": "Cork"
        }
      ],
      "rowCount": 3
    },
    {
      "pivotDimensionHeaders": [
        {
          "value": "United States"
        },
        {
          "value": "United Kingdom"
        },
        {
          "value": "India"
        }
      ],
      "rowCount": 3
    },
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            {
              "value": "Direct"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Organic Search"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "Paid Social"
            }
          ]
        },
        {
          "dimensionValues": [
            {
              "value": "(other)"
            }
          ]
        }
      ],
      "rowCount": 4
    }
  ],
```

</details>

<details>
<summary>But the expectation is for the following shape:</summary>

```json
  "pivotHeaders": [
    {
      "pivotDimensionHeaders": [
        { "dimensionValues": [{ "value": "properties/12345/audiences/1" }] },
        { "dimensionValues": [{ "value": "properties/12345/audiences/2" }] }
      ],
      "rowCount": 2
    },
    {
      "pivotDimensionHeaders": [
        {
          "dimensionValues": [
            { "value": "Dublin" },
            { "value": "United States" }
          ]
        },
        {
          "dimensionValues": [
            { "value": "(not set)" },
            { "value": "United Kingdom" }
          ]
        },
        { "dimensionValues": [{ "value": "Cork" }, { "value": "India" }] }
      ],
      "rowCount": 6
    },
    {
      "pivotDimensionHeaders": [
        { "dimensionValues": [{ "value": "Direct" }] },
        { "dimensionValues": [{ "value": "Organic Search" }] },
        { "dimensionValues": [{ "value": "Paid Social" }] },
        { "dimensionValues": [{ "value": "(other)" }] }
      ],
      "rowCount": 4
    }
  ],
```

</details>

I took a look and have found a fix - in the interest of time I have pushed the change directly (see commit 82463974f19634530512bb6b3636bbdda012e23d). Please review it and if you are happy with the change, update the fixture file accordingly.

File: assets/js/modules/analytics-4/components/audience-segmentation/dashboard/utils.js
Comment (techanvil) [inline assets/js/modules/analytics-4/components/audience-segmentation/dashboard/utils.js:1]: These utilities are specific to the `AudienceTile` component and this file can be moved to the `AudienceTile` folder.

**Review by techanvil — CHANGES_REQUESTED** [review]
Thanks @ankitrox, just a few last comments to address and this should be good to go.

**Review by techanvil — APPROVED** [review]
LGTM, nice work @ankitrox!

**Pull request merged**